modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileSender.java

```diff
@@ -0,0 +1,227 @@
+                processTransferWithNextAsync(transfer);
+            } else {
+                requests.add(transfer);
+            }
+        }
+    }
+
+    /**
+     * Process the given transfer and then process the next transfer if the rate limiter is available and there is a next transfer.
+     *
+     * @return A future that will be completed when the request is processed.
+     */
+    private CompletableFuture<Void> processTransferWithNextAsync(FileTransfer transfer) {
+        return sendTransfer(transfer)
+                .thenComposeAsync(v -> {
+                    synchronized (rateLimiter) {
+                        FileTransfer nextTransfer = requests.poll();
+
+                        // If there is a next transfer, process it.
+                        // Otherwise, release the rate limiter.
+                        if (nextTransfer != null) {
+                            return processTransferWithNextAsync(nextTransfer);
+                        } else {
+                            rateLimiter.release();
+                            return completedFuture(null);
+                        }
+                    }
+                }, executorService);
+    }
+
+    /**
+     * Sends the file to the node with the given consistent id.
+     *
+     * @param transfer The transfer to send.
+     * @return A future that will be completed when the file is sent. The future will be completed successfully always, even if the file is
+     *         not sent.
+     */
+    private CompletableFuture<Void> sendTransfer(FileTransfer transfer) {
+        return sendFile(transfer.receiverConsistentId, transfer.transferId, transfer.path, transfer.shouldBeCancelled)
+                .handle((v, e) -> {
+                    if (e == null) {
+                        transfer.result.complete(null);
+                    } else {
+                        transfer.result.completeExceptionally(e);
+                    }
+                    return null;
+                });
+    }
+
+    /**
+     * Sends the file to the node with the given consistent id.
+     *
+     * @param receiverConsistentId The consistent id of the node to send the file to.
+     * @param id The id of the file transfer.
+     * @param path The path of the file to send.
+     * @return A future that will be completed when the file is sent.
+     */
+    private CompletableFuture<Void> sendFile(String receiverConsistentId, UUID id, Path path, AtomicBoolean shouldBeCancelled) {
+        if (path.toFile().length() == 0) {
+            return completedFuture(null);
+        } else {
+            return supplyAsync(() -> {
+                try {
+                    return FileChunkMessagesStream.fromPath(chunkSize, id, path);
+                } catch (IOException e) {
+                    throw new FileTransferException("Failed to create a file transfer stream", e);
+                }
+            }, executorService)
+                    .thenCompose(stream -> {
```

Comment:
What's the deal with "thenCompose"? You could put this code right into "supplyAsync"?
Please explain, don't just close the conversation.